File: codewars/crack-the-encrypted-email/kata.rb
class String
  def rot(num)
    self.split("").collect { |ch|
      if ch =~ /^[a-z]$/
	      ((ch.ord + num - 'a'.ord) % 26 + 'a'.ord).chr
      else
	      ch
      end
    }.join("")
  end
end

def crack(text)
  26.times { |num|
    solution = text.rot(num)
    return solution if solution =~ /test/ or solution =~ /price/
  }
end

secret = "ikhcxvmutgdknimykxlahluxzbgvhffngbvtmbhgehpxkzxetmhikbvxlurmabkmrixkvxgmhgmnxlwtrtgwkxwnvxfbedlatdxikbvxlurtmextlmhgxwheetk"
p crack(secret)

p crack('jxyijuijsqiufqiiut')
p crack('btejeuijtuftu')
p crack('naqguvfgrfg')

File: codewars/crack-the-encrypted-email/kata.js
// code
String.prototype.rotChar = function(num, position) {
  const aOrd = "a".charCodeAt(0);
  const zOrd = "z".charCodeAt(0);
  const alphabetLength = zOrd - aOrd + 1;
  const charCode = this[position].charCodeAt(0);
  const rotatedCharCode = (charCode - aOrd + num) % alphabetLength + aOrd;
  return String.fromCharCode(rotatedCharCode);
};

String.prototype.rot = function(num) {
  return this.split("").map((char, position) => {
    if (/[a-z]/.test(char)) return this.rotChar(num, position);
    return char;
  }).join("");
};

const crack = text => {
  const limit = 26;
  for (let i = 1; i <= limit; i++) {
    let solution = text.rot(i);
    if (/price|test/.test(solution)) return solution;
  }
  throw new Error('No solution found!');
};

// test
console.log(crack('ikhcxvmutgdknimykxlahluxzbgvhffngbvtmbhgehpxkzxetmhikbvxlurmabkmrixkvxgmhgmnxlwtrtgwkxwnvxfbedlatdxikbvxlurtmextlmhgxwheetk'));
console.log(crack('jxyijuijsqiufqiiut'));
console.log(crack('btejeuijtuftu'));
console.log(crack('naqguvfgrfg'));

